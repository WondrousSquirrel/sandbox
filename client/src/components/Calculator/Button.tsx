import React from 'react';

interface Button {
    text: string;
    backgroundColor: string;
    className?: string
};

const Button = (props: Button) => {
    return <div className={'button ' + props.className} style={{backgroundColor: props.backgroundColor }}>
        <p className='text'>{props.text}</p>
    </div>;
};

export default Button;