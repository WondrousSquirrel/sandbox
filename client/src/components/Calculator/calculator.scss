.calculator {
    display: flex;
    flex-direction: column;
    width: 300px;
    padding: 10px 10px;
    height: 480px;
    border: 1px solid #DCDEE5;
}

.display {
    display: flex;
    align-items: flex-end;
    flex-direction: column-reverse;
    margin-top: 10px;
    padding: 20px 10px;
    width: 280px;
}

.wrapper {
    display: flex;
    justify-content: space-around;
}

.grid {
    display: grid;
    grid-template-columns: 25% 25% 25% 25%;
    grid-template-rows: 20% 20% 20% 20% 20%;
}

.bottom-grid {
    display: grid;
    grid-template-columns: 50% 25% 25%;
    grid-template-areas: "double-sized dot-btn equal-btn";
}

.double-sized {
    grid-area: double-sized;
}
.dot-btn {
    grid-area: dot-btn;
}
.equal-btn {
    grid-area: equal-btn;
}

.button {
    text-align: center;
    padding: 20px 0;
    margin: 5px;
    border-radius: 5%;
    cursor: pointer;
}


.button:hover {
    box-shadow: 10px 10px 23px -8px #6A7FDB;
    -webkit-box-shadow: 10px 10px 23px -8px #6A7FDB;
    -moz-box-shadow: 10px 10px 23px -8px #6A7FDB;
}

.button:active {
    background-color: #FFD5AD !important;
    color: black;
    box-shadow: none;
    -webkit-box-shadow: none;
    -moz-box-shadow:  none

}

@media screen and (max-height: 700px) {
    .calculator {
        width: 300px;
        height: 400px;
    }
    .button {
        text-align: center;
        padding: 15px 0;
        margin: 3px;
        border-radius: 5%;
    }
  }
