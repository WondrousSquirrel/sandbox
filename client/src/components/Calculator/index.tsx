import React, { useContext, FunctionComponent, useState } from 'react';
import ThemeContext from '../theme';

import './calculator.scss'
import Button from './Button';

type Operator = '+' | '-' | '*' | '/';

const Calculator: FunctionComponent = () => {

    const theme = useContext(ThemeContext);
    const [display, setDisplay] = useState<string>('0') // правая часть выражения
    const [memory, setMemory] = useState<number>(0)
    // не позволяет установить оператор с левой стороны операнда
    const [waitOperand, setWaitForOperand] = useState<boolean>(true) 
    // записиываем выбранный оператор и выполняем указанную операцию
    const [pendingOperator, setPendingOperator] = useState<Operator>()
    const [result, setResult] = useState<number>(0) // ялвяется левой частью выражения
    // переключаем операции на работу с процентами
    const [percentMode, setPercentMode] = useState<boolean>(false);
    const percentage = () => {
        setPercentMode(!percentMode);
    }

    const calculate = (rightOperand: number, pendingOperator: Operator): boolean => {
        let newResult = result;
        let convertToPercent = result / 100 * rightOperand;
        // выбераем необходимую операцию
        switch (pendingOperator) {
            case '+':
                // проверка на работу с процентами
                percentMode ? newResult +=  convertToPercent: newResult += rightOperand
                break
            case '-':
                percentMode ? newResult -= convertToPercent : newResult -= rightOperand
                break
            case '*':
                percentMode ? newResult *= convertToPercent : newResult *= rightOperand
                break
            case '/':
                if (rightOperand === 0) {
                    return false
                }
                percentMode ? newResult /= convertToPercent : newResult /= rightOperand
                break;
        }

        console.log('before setting ' + newResult)
        setResult(newResult)
        setDisplay(newResult.toString().slice(0, 15))
        return true
    }

    const addNumber = (digit: Number) => {
        // ставим цифру
        let newDisplay = display

        // если экран пуст и выбрано число 0 или длина более 15 символов - ничего не делаем
        if ((display === '0' && digit === 0) || display.length > 15) {
          return
        }
        // после добавления числа позволяем выбирать операцию
        if (waitOperand) {
          newDisplay = ''
          setWaitForOperand(false)
        }
    
        if (display !== '0') {
          newDisplay = newDisplay + digit.toString()
        } else {
          newDisplay = digit.toString()
        }
    
        setDisplay(newDisplay)
    }


    const letDot = () => {
        let newDisplay = display

        if (waitOperand) {
            newDisplay = '0'
        }

        if (newDisplay.indexOf('.') === -1) {
            newDisplay = newDisplay + '.'
        }

        setDisplay(newDisplay)
        setWaitForOperand(false)
    }


    const addOperator = (operator: Operator) => {
        const operand = Number(display)

        // если выбран оператор выполняем операцию
        if (typeof pendingOperator !== 'undefined' && !waitOperand) {
            if (!calculate(operand, pendingOperator)) { // в случае ошибки выходим из функции
                return
            }
        } else { // или записываем число в левую часть выражения
            setResult(operand)
        }

        setPendingOperator(operator) // записываем опирацию
        setWaitForOperand(true)
    }



    const changeSign = () => {
        // меняем знак числа
        const value = Number(display)

        if (value > 0) {
            setDisplay('-' + display)
        } else if (value < 0) {
            setDisplay(display.slice(1)) // отсекаем перый индекс правого выражения
        }
    }

    const equal = () => {
        const operand = Number(display)

        if (typeof pendingOperator !== 'undefined' && !waitOperand) {
            if (!calculate(operand, pendingOperator)) {
                return
            }

            setPendingOperator(undefined)
        } else {
            setDisplay(operand.toString())
        }

        setResult(operand)
        setWaitForOperand(true)
        console.log('result ' + result)
    }


    const allClear = () => {
        // чистка всех данных
        setMemory(0)
        setResult(0)
        setPendingOperator(undefined)
        setDisplay('0')
        setWaitForOperand(true)
    }

    const memoryRecall = () => {
        // получаем значение из памяти
        setDisplay(memory.toString())
        setWaitForOperand(true)
    }

    const memoryClear = () => {
        setMemory(0)
        setWaitForOperand(true)
    }

    const addToMemory = () => {
        setMemory(memory + Number(display))
        setWaitForOperand(true)
    }

    const remooveFromMemory = () => {
        setMemory(memory - Number(display))
        setWaitForOperand(true)
    }

    return (
        <div className='centered'>

            <div className='calculator' style={{ backgroundColor: theme.secondColor, borderColor: theme.borderColor }}>
                <div className='display'>
                    <div className='memory' style={{ color: theme.color2 }}>
                        <p>{ percentMode ? '%': null}</p>
                        <p>{memory === 0 ? null : memory}</p>
                    </div>
                    <div className='input'>
                        <span>{display.length > 0 ? display : 0}</span>
                    </div>
                </div>
                <div className="grid">
                    <Button text='AC' backgroundColor={theme.buttonColor} onClick={allClear} />
                    <Button text='+/-' backgroundColor={theme.buttonColor} onClick={changeSign} />
                    <Button text='%' backgroundColor={theme.buttonColor} onClick={percentage} />
                    <Button text='/' backgroundColor={theme.buttonColor2} onClick={() => addOperator('/')} />
                    <Button text='mc' backgroundColor={theme.buttonColor3} onClick={memoryClear} />
                    <Button text='mr' backgroundColor={theme.buttonColor3} onClick={memoryRecall} />
                    <Button text='m-' backgroundColor={theme.buttonColor3} onClick={remooveFromMemory} />
                    <Button text='m+' backgroundColor={theme.buttonColor2} onClick={addToMemory} />
                    <Button text='7' backgroundColor={theme.buttonColor3} onClick={() => addNumber(7)} />
                    <Button text='8' backgroundColor={theme.buttonColor3} onClick={() => addNumber(8)} />
                    <Button text='9' backgroundColor={theme.buttonColor3} onClick={() => addNumber(9)} />
                    <Button text='*' backgroundColor={theme.buttonColor2} onClick={() => addOperator('*')} />
                    <Button text='4' backgroundColor={theme.buttonColor3} onClick={() => addNumber(4)} />
                    <Button text='5' backgroundColor={theme.buttonColor3} onClick={() => addNumber(5)} />
                    <Button text='6' backgroundColor={theme.buttonColor3} onClick={() => addNumber(6)} />
                    <Button text='-' backgroundColor={theme.buttonColor2} onClick={() => addOperator('-')} />
                    <Button text='1' backgroundColor={theme.buttonColor3} onClick={() => addNumber(1)} />
                    <Button text='2' backgroundColor={theme.buttonColor3} onClick={() => addNumber(2)} />
                    <Button text='3' backgroundColor={theme.buttonColor3} onClick={() => addNumber(3)} />
                    <Button text='+' backgroundColor={theme.buttonColor2} onClick={() => addOperator('+')} />
                </div>
                <div className="bottom-grid">
                    <Button text='0' backgroundColor={theme.buttonColor3} className='double-sized' onClick={() => addNumber(0)} />
                    <Button text=',' backgroundColor={theme.buttonColor3} className='dot-btn' onClick={letDot} />
                    <Button text='=' backgroundColor={theme.buttonColor2} className='equal-btn' onClick={equal} />

                </div>
            </div>

        </div>
    )
}

export default Calculator;