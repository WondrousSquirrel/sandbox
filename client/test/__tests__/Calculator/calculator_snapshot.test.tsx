import React from 'react';
import renderer from 'react-test-renderer';

import Calculator from '../../../src/components/Calculator';
import Button from '../../../src/components/Calculator/Button';

const btnProps = {
    text: '1',
    backgroundColor: '#DBE0E6',
    onClick: () => { }
}

describe('Snapshot Testing', () => {
    test('calculator snapshot', () => {
        const tree = renderer.create(<Calculator />).toJSON();
        expect(tree).toMatchSnapshot();
    })

    test('button snapshot', () => {
        const tree = renderer.create(
        <Button 
            text={btnProps.text}
            backgroundColor={btnProps.backgroundColor}
            onClick={btnProps.onClick} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })

})


