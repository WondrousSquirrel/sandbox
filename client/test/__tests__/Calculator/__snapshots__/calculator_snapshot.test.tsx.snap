// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing button snapshot 1`] = `
<div
  className="button undefined"
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "#DBE0E6",
    }
  }
>
  <p
    className="text"
  >
    1
  </p>
</div>
`;

exports[`Snapshot Testing calculator snapshot 1`] = `
<div
  className="centered"
>
  <div
    className="calculator"
    style={
      Object {
        "backgroundColor": "#F5F5F5",
        "borderColor": "#EDE8EB",
      }
    }
  >
    <div
      className="display"
    >
      <div
        className="memory"
        style={
          Object {
            "color": "#24203880",
          }
        }
      >
        <p />
        <p />
      </div>
      <div
        className="input"
      >
        <span>
          0
        </span>
      </div>
    </div>
    <div
      className="grid"
    >
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#DBE0E6",
          }
        }
      >
        <p
          className="text"
        >
          AC
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#DBE0E6",
          }
        }
      >
        <p
          className="text"
        >
          +/-
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#DBE0E6",
          }
        }
      >
        <p
          className="text"
        >
          %
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#FA9961",
          }
        }
      >
        <p
          className="text"
        >
          /
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          mc
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          mr
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          m-
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#FA9961",
          }
        }
      >
        <p
          className="text"
        >
          m+
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          7
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          8
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          9
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#FA9961",
          }
        }
      >
        <p
          className="text"
        >
          *
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          4
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          5
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          6
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#FA9961",
          }
        }
      >
        <p
          className="text"
        >
          -
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          1
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          2
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          3
        </p>
      </div>
      <div
        className="button undefined"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#FA9961",
          }
        }
      >
        <p
          className="text"
        >
          +
        </p>
      </div>
    </div>
    <div
      className="bottom-grid"
    >
      <div
        className="button double-sized"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          0
        </p>
      </div>
      <div
        className="button dot-btn"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#C2C2C2",
          }
        }
      >
        <p
          className="text"
        >
          ,
        </p>
      </div>
      <div
        className="button equal-btn"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#FA9961",
          }
        }
      >
        <p
          className="text"
        >
          =
        </p>
      </div>
    </div>
  </div>
</div>
`;
